File: Api/algoritmia_api/tables/contests.py
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, HTTPException, Query, Depends
from pydantic import BaseModel, AnyUrl, Field

from .. import db
from .auth import get_current_user

router = APIRouter(prefix="/contests", tags=["Contests"])

DDL = """
CREATE TABLE IF NOT EXISTS contests (
    id BIGSERIAL PRIMARY KEY,
    title TEXT NOT NULL,
    platform TEXT NOT NULL,
    url TEXT NOT NULL,
    tags TEXT[],
    difficulty INT CHECK (difficulty BETWEEN 1 AND 5),
    added_by BIGINT REFERENCES users(id) ON DELETE SET NULL,
    format TEXT NOT NULL,
    start_at TIMESTAMPTZ NOT NULL,
    end_at TIMESTAMPTZ NOT NULL,
    location TEXT,
    season TEXT,
    notes TEXT
);
"""


def ensure_table(conn) -> None:
    with conn.cursor() as cur:
        cur.execute(DDL)
    conn.commit()


def row_to_contest(row: dict) -> dict:
    """Normalize DB row → JSON shape expected by the frontend."""
    return {
        "id": row["id"],
        "title": row["title"],
        "platform": row["platform"],
        "url": row["url"],
        "tags": row["tags"] or [],
        "difficulty": row["difficulty"] or 3,  # default if NULL
        "format": row["format"],
        "startsAt": row["start_at"].isoformat(),
        "endsAt": row["end_at"].isoformat(),
        "location": row["location"] or "",
        "season": row["season"] or "",
        "notes": row["notes"] or "",
    }

class ContestCreate(BaseModel):
    title: str
    platform: str
    url: AnyUrl
    tags: Optional[list[str]] = None
    difficulty: Optional[int] = Field(default=None, ge=1, le=5)
    format: str
    start_at: datetime
    end_at: datetime
    location: Optional[str] = None
    season: Optional[str] = None
    notes: Optional[str] = None


class ContestUpdate(BaseModel):
    title: Optional[str] = None
    platform: Optional[str] = None
    url: Optional[AnyUrl] = None
    tags: Optional[list[str]] = None
    difficulty: Optional[int] = Field(default=None, ge=1, le=5)
    format: Optional[str] = None
    start_at: Optional[datetime] = None
    end_at: Optional[datetime] = None
    location: Optional[str] = None
    season: Optional[str] = None
    notes: Optional[str] = None


@router.get("")
def list_contests(
    platform: Optional[str] = None,
    season: Optional[str] = None,
    upcoming_only: bool = Query(False),
):
    base = "SELECT * FROM contests"
    clauses = []
    params: list = []

    if platform:
        clauses.append("platform=%s")
        params.append(platform)
    if season:
        clauses.append("season=%s")
        params.append(season)
    if upcoming_only:
        clauses.append("end_at >= NOW()")

    if clauses:
        base += " WHERE " + " AND ".join(clauses)

    base += " ORDER BY start_at DESC LIMIT 200"

    with db.connect() as conn:
        rows = db.fetchall(conn, base, params)
    return {"items": [row_to_contest(r) for r in rows]}


@router.get("/{contest_id}")
def get_contest(contest_id: int):
    with db.connect() as conn:
        row = db.fetchone(conn, "SELECT * FROM contests WHERE id=%s", [contest_id])
    if not row:
        raise HTTPException(status_code=404, detail="Contest not found")
    return row_to_contest(row)


@router.post("")
def create_contest(payload: ContestCreate, auth=Depends(get_current_user)):
    """
    Only 'coach' or 'admin' can create contests.
    `auth` looks like: {"session": ..., "user": ...}
    """
    user = auth["user"]
    role = user.get("role")
    user_id = user["id"]

    if role not in ("coach", "admin"):
        raise HTTPException(status_code=403, detail="Only coaches/admins can create contests")

    if payload.start_at >= payload.end_at:
        raise HTTPException(status_code=400, detail="start_at must be before end_at")

    with db.connect() as conn:
        row = db.fetchone(
            conn,
            """
            INSERT INTO contests(
                title, platform, url, tags, difficulty,
                added_by, format, start_at, end_at, location, season, notes
            )
            VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
            RETURNING *
            """,
            [
                payload.title,
                payload.platform,
                str(payload.url),
                payload.tags,
                payload.difficulty,
                user_id,
                payload.format,
                payload.start_at,
                payload.end_at,
                payload.location,
                payload.season,
                payload.notes,
            ],
        )
    return row_to_contest(row)


@router.patch("/{contest_id}")
def update_contest(contest_id: int, payload: ContestUpdate, auth=Depends(get_current_user)):
    user = auth["user"]
    role = user.get("role")

    if role not in ("coach", "admin"):
        raise HTTPException(status_code=403, detail="Only coaches/admins can update contests")

    data = payload.model_dump(exclude_unset=True)

    if "url" in data and data["url"] is not None:
        data["url"] = str(data["url"])

    if not data:
        raise HTTPException(status_code=400, detail="No fields to update")

    # optional: validate start_at < end_at if both present
    if "start_at" in data and "end_at" in data:
        if data["start_at"] >= data["end_at"]:
            raise HTTPException(status_code=400, detail="start_at must be before end_at")

    cols = ", ".join(f"{k}=%s" for k in data.keys())
    params = list(data.values()) + [contest_id]

    with db.connect() as conn:
        row = db.fetchone(conn, f"UPDATE contests SET {cols} WHERE id=%s RETURNING *", params)

    if not row:
        raise HTTPException(status_code=404, detail="Contest not found")
    return row_to_contest(row)


@router.delete("/{contest_id}")
def delete_contest(contest_id: int, auth=Depends(get_current_user)):
    user = auth["user"]
    role = user.get("role")

    if role not in ("coach", "admin"):
        raise HTTPException(status_code=403, detail="Only coaches/admins can delete contests")

    with db.connect() as conn:
        count = db.execute(conn, "DELETE FROM contests WHERE id=%s", [contest_id])

    if count == 0:
        raise HTTPException(status_code=404, detail="Contest not found")
    return {"deleted": True}

File: src/lib/contests.ts
import { postJSON } from './api';
import { API_BASE, HttpError } from '@/lib/api';
import type { Platform, ContestFormat, Difficulty, Contest } from './types';

export type CreateContestPayload = {
  title: string;
  platform: Platform;
  url: string;
  tags: string[];
  difficulty: Difficulty;
  format: ContestFormat;
  start_at: string; // ISO string
  end_at: string;   // ISO string
  location: string;
  season: string;
  notes: string;
};

type ApiContestRow = {
  id: number;
  title: string;
  platform: string;
  url: string;
  tags: string[] | null;
  difficulty: number | null;
  added_by: number | null;
  format: string;
  start_at: string;
  end_at: string;
  location: string | null;
  season: string | null;
  notes: string | null;
};

function toUiContest(row: ApiContestRow): Contest {
  return {
    id: String(row.id),
    title: row.title,
    platform: row.platform as Platform,
    url: row.url,
    tags: row.tags ?? [],
    difficulty: (row.difficulty ?? 3) as Difficulty,
    format: row.format as ContestFormat,
    startsAt: row.start_at,
    endsAt: row.end_at,
    location: row.location ?? '',
    season: row.season ?? '',
  };
}

export async function createContest(payload: CreateContestPayload): Promise<Contest> {
  const row = await postJSON<ApiContestRow>('/contests', payload);
  return toUiContest(row);
}

export async function updateContest(
  contestId: string,
  data: {
    title?: string;
    platform?: Platform;
    url?: string;
    tags?: string[];
    difficulty?: Difficulty;
    format?: ContestFormat;
    start_at?: string; // ISO
    end_at?: string;   // ISO
    location?: string;
    season?: string;
    notes?: string;
  }
): Promise<Contest> {
  const res = await fetch(`${API_BASE}/contests/${contestId}`, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    credentials: 'include', // required for session cookie
    body: JSON.stringify(data),
  });

  if (!res.ok) {
    let detail = 'No se pudo actualizar el concurso.';
    let errorBody: unknown = null;

    try {
      const json: unknown = await res.json();
      errorBody = json;
      if (json && typeof json === 'object' && 'detail' in json && typeof (json as any).detail === 'string') {
        detail = (json as any).detail;
      }
    } catch {
      // ignore JSON parse errors
    }

    throw new HttpError(res.status, detail, errorBody);
  }

  const json = (await res.json()) as ApiContestRow;
  return toUiContest(json);
}

export async function deleteContest(contestId: string): Promise<{ deleted: boolean }> {
  const res = await fetch(`${API_BASE}/contests/${contestId}`, {
    method: 'DELETE',
    credentials: 'include',
  });

  if (!res.ok) {
    let detail = 'No se pudo eliminar el concurso.';
    let errorBody: unknown = null;

    try {
      const json: unknown = await res.json();
      errorBody = json;
      if (json && typeof json === 'object' && 'detail' in json && typeof (json as any).detail === 'string') {
        detail = (json as any).detail;
      }
    } catch {
      // ignore JSON parse errors
    }

    throw new HttpError(res.status, detail, errorBody);
  }

  return res.json(); // e.g. { deleted: true }
}
